### PR TITLE
AB#78 improve code editing experience

### PR DESCRIPTION
**Pull request by sir-marc**

### CODE REVIEW

File: game-frontend/src/components/user-script/monaco-helper.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  return monacoInstance;
+};
+
+export const useSaveHotKey = (editorRef: EditorRef, save: () => void): void => {
+  const monacoInstance = useMonacoInstance();
+
+  useEffect(() => {
+    if (editorRef && editorRef.current && monacoInstance) {
+      const m = monacoInstance as MonacoInstance;
+      editorRef.current.addAction({
```
Comment (StuckiSimon) [inline game-frontend/src/components/user-script/monaco-helper.ts]: In the current implementation every code change in the editor will retrigger `addAction`.
It does work, but it's certainly not the proper way to handle this kind of behavior.
Consider refactoring it to reduce the number of unnecessary calls on a foreign API?

Comment (michaelschufi) [reply game-frontend/src/components/user-script/monaco-helper.ts]: This is resolved here: https://github.com/code-and-conquer/darwin/pull/56

File: game-frontend/src/components/user-script/index.tsx [hunk 2]
```diff
@@ -1,36 +1,42 @@
...
 
-    const scriptUpdate: ScriptUpdate = {
-      type: 'scriptUpdate',
-      payload: {
-        script: userScript,
-      },
-    };
-
-    send(scriptUpdate);
+  const onChange: ControlledEditorOnChange = (_ev, value) => {
```
Comment (StuckiSimon) [inline game-frontend/src/components/user-script/index.tsx]: Didn't we use the convention of a simple `_` for omitted params?

File: game-frontend/src/components/user-script/monaco-helper.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  editorRef: EditorRef,
+  save: (value: string) => void
+): void => {
+  const monacoInstance = useMonacoInstance();
+  const SAVE_ACTION_ID = 'DARWIN_SAVE';
+
+  useEffect(() => {
+    if (editorRef && editorRef.current && monacoInstance) {
+      const editorSaveAction = editorRef.current.getAction(SAVE_ACTION_ID);
+      if (!editorSaveAction || editorSaveAction.run !== save) {
```
Comment (StuckiSimon) [inline game-frontend/src/components/user-script/monaco-helper.ts]: What's the idea behind the `editorSaveAction.run !== save`?

Comment (sir-marc) [reply game-frontend/src/components/user-script/monaco-helper.ts]: To change the editorsAction when the save method changes

Comment (StuckiSimon) [reply game-frontend/src/components/user-script/monaco-helper.ts]: And how exactly does this work when you assign a new function to the run property of the action? (See line 45)
```
run: editor => {
  save(editor.getValue());
},
```

And therefore `editorSaveAction.run` will **never** be `save`. (If you'd have assigned `save` to `run:`, yeah that would work - and that's probably what you meant to do?)
Or am I missing something?

Comment (sir-marc) [reply game-frontend/src/components/user-script/monaco-helper.ts]: Yeah, you're right, that's what i meant to do

**Review by michaelschufi — APPROVED** [review]

**Review by StuckiSimon — APPROVED** [review]
As discussed: we might refactor the save callback handling to make it easier to comprehend - but the rest looks good 👍

**Pull request merged**